fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/DefaultOCFLObjectSession.java

```diff
@@ -89,85 +114,267 @@ public void write(final String subpath, final InputStream stream) {
+                return new FileInputStream(stagedPath.toFile());
+            } catch (final FileNotFoundException e) {
+                throw new PersistentItemNotFoundException(format("Could not find %s within object %s",
+                        subpath, objectIdentifier));
+            }
+        } else if (!objectDeleted) {
+            // Fall back to the head version
+            return readVersion(subpath, ObjectVersionId.head(objectIdentifier));
         }
-        return null;
-    }
 
-    private boolean hasStagedChanges(final String subpath) {
-        return false;
+        throw new PersistentItemNotFoundException(format("Could not find %s within object %s",
+                subpath, objectIdentifier));
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
-    public InputStream read(final String subpath, final String version) {
-        final AtomicReference<InputStream> contentRef = new AtomicReference<>();
+    public InputStream read(final String subpath, final String version) throws PersistentItemNotFoundException {
+        // If the object was deleted, then only uncommitted staged files can be available
+        if (objectDeleted) {
+            throw new PersistentItemNotFoundException(format("Could not find %s within object %s",
+                    subpath, objectIdentifier));
+        }
 
-        ocflRepository.readObject(ObjectVersionId.version(objectIdentifier, version), reader -> {
-            contentRef.set(reader.getFile(subpath));
-        });
+        return readVersion(subpath, ObjectVersionId.version(objectIdentifier, version));
+    }
 
-        return contentRef.get();
+    private InputStream readVersion(final String subpath, final ObjectVersionId version) throws PersistentItemNotFoundException {
+        try {
+            // read the head version of the file from the ocfl object
+            final var file = ocflRepository.getObject(version)
+                    .getFile(subpath);
+            if (file == null) {
+                throw new PersistentItemNotFoundException(format("Could not find %s within object %s version %s",
+                        subpath, objectIdentifier, version.getVersionId()));
+            }
+            return file.getStream();
+        } catch (final NotFoundException e) {
+            throw new PersistentItemNotFoundException(format(
+                    "Unable to read %s from object %s version %s, object was not found.",
+                    subpath, objectIdentifier, version.getVersionId()));
+        }
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
-    public void prepare() {
-        // TODO check for conflicts
+    public synchronized void prepare() {
+        // TODO check for conflicts and lock the object
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
-    public void commit(final CommitOption commitOption) {
+    public synchronized String commit(final CommitOption commitOption) throws PersistentStorageException {
```

Comment:
Is there any concurrency control above this class? If not, is it a problem if any of the methods on this class are executed concurrently? `synchronized` only helps prevent concurrent execution of the same method. If that doesn't matter here, that's fine, but, if it does matter, you'd probably need to use a lock.

Reply:
@pwinckles My understanding (I had to remind myself about the specifics earlier when working on this) is that a synchronized method is the equivalent of wrapping the body of the method with `synchronized (this) { }`, which would cause blocking on calls to any other synchronized methods. 

My intent was to prevent simultaneous write operations to the same object/staging path, but to allow simultaneous reads to continue to happen. Perhaps that's not realistic though, since a file could get deleted/overwritten mid-read. If that's the case, then I should probably synchronize the read methods as well, or more precisely lock based on stagingPath and/or ocflRepository usage. Probably the latter. I was hoping to keep from having synchronize or locking blocks all over the place, but probably not avoidable.

I don't think a higher level will manage synchronized access to the whole class.

Reply:
You're absolutely right about `synchronized`. Sorry!

Yes, there is a danger that a resource could be deleted while it's being read. However, using a lock to protect reads would be quite tricky because you're using InputStreams. If you want to guarantee a read of a resource in Fedora staging you'd have to hold the read lock until the stream is closed.

Reply:
Huh, I got a kind of surprisingly result when testing to see what would happen if I deleted the file after opening an input stream. It continued to read the contents:
```
        final File dme = tempFolder.newFile("deleteme.txt");
        FileUtils.writeStringToFile(dme, "whatever");

        final InputStream inne = new FileInputStream(dme);
        inne.read(new byte[4]);
        dme.delete();
        System.out.println(dme.exists());
        System.out.print(IOUtils.toString(inne));
```
outputs:
```
false
ever
```
Moving the file produced the same result. So maybe I don't need to worry about the file going away? I get the same result with a 10000 character string. Either way you make a good point about the locking not doing much since i'm passing back an inputstream that would be consumed after the lock was released. My inclination is to leave the read methods unsynchronized. It might be a good idea to introduce a concurrency test at some point, although i'm not sure how to structure it at the moment for this class

Reply:
Interesting... Does the same thing happen when you do that to a file created outside the test? I wonder if Java is fudging anything. Like saying something is deleted when it's only queued for deletion or buffering file contents in memory.

Reply:
just tried creating the file outside of java and then doing the same operations (minus the write portion). Same result, says the file is deleted, outputs file contents, and the file is gone when i look afterwards. Maybe it reads the whole thing? or maybe it has a reference to the file, and the OS doesn't truly delete its contents until all the references are cleared.

Reply:
Cool!